proposals/20210818-driver-semver.md

```diff
@@ -0,0 +1,95 @@
+to get the probe version (`PPM_IOCTL_GET_PROBE_VERSION`) but it is not used anywhere.
+Again, what is versioned is the libscap consumer, not actual libscap.
+
+## Motivation
+
+Introducing a machine-usable API versioning scheme will let us:
+
+1. Cut down the number of probes needed to be built. Instead of a probe for each
+(kernel version, consumer name, consumer version) tuple, we would only need one
+for each (kernel version, driver API version). Given the relatively slow development
+on the kernel side, a single driver API version might live for a long time
+
+2. Make upgrades easier. Currently, the probe has to be unloaded and a new one loaded
+in its place. With the API versioning scheme in place, usually there won't even
+be a different probe. If there is one, it will usually be forwards-compatible,
+even though it might miss some bug fixes. Only if the versions are truly incompatible,
+an unload/reload will be required. Note that the API version can live in the module
+*name* itself, which would let us have multiple versions loaded at the same time,
+if we decide to go that way.
+
+3. Ship the probes in Linux distributions. With a single consumer-agnostic probe
+per kernel, it becomes realistic for Linux distributions to ship a prebuilt probe
+for their kernels. That would make all libscap consumers work out of the box,
+without worrying about shipping the probe.
+
+4. Support older consumer versions with new kernels. Whenever a new kernel comes out,
+probes need to be built to support it. These probes currently cannot be used
+for older consumer releases, even though there are no technical issues that would
+prevent it.
+
+## Goals
+
+* Make the probes reusable across libscap consumers and their versions
+
+## Non-goals
+
+* Make the probes reusable across kernel versions (this is probably impossible
+  in a general way for the kernel module, but BTF/CO-RE may help for the eBPF
+  probe).
+
+## The plan
+
+1. Introduce an API version embedded in the userspace and kernel code
+  * The version number will be a single 64-bit number that can be decomposed
+    to the three semantic versioning components (major, minor, patch)
+
+2. Extend the review process to ensure the API version is incremented when needed.
+  Note that e.g. adding support for a new kernel should not result in an API
+  version increase (if the probe failed to build for that particular kernel
+  before).
+
+3. Verify the API version of the kernel module/eBPF probe when starting
+  the libscap consumer:
+  * different major versions cause a hard error
+  * kernel minor < userspace minor causes a hard error
+  * kernel minor == userspace minor and kernel patch < userspace patch causes
+    a warning (the probe is compatible but has known bugs, fixed in later
+    versions)
+
+4. Remove consumer name and version from the libscap build process
+  * The probe should always be named `scap_probe` and use the API version to identify
+    itself
```

Comment:
This proposal addresses a lot of usability and maintenance issues for API consumers. I like the idea of a default/canonical name (`scap`). 

I lean towards a parametric build, where the consumer name can be passed to override a default consumer name.

Reply:
I'm fine with leaving an option to specify the module name as an internal I-know-what-I'm-doing hack but IMO it shouldn't be the default way of building. I'll try to update the wording to better express the idea.

Reply:
This is a nice proposal that is really needed to support libsinsp and libscap as separate entities from their consumers.   The probe should definitely come with a different consumer agnostic name, and that name and version should be used in the pre-built probes that are downloadable to make maintenance easier.  

As for whether the name should be parameterized, I don't have a strong opinion.  I think most consumers are going to want to take advantage of the pre-built probes they can download; however, there may be cases where two different consumers want to have two different probes.  As such, I think probably making the name configurable but defaulted, wouldn't add a lot of complexity to the build process, but give flexibility to change for those who really want too.  

Finally, one thing that I think might go really well with this proposal would be to:
1.) move a version of the falco-driver-loader and docker-entrypoint.sh script into the libraries, as these are  key scripts for any consumer using the probe.

Reply:
Moving `falco-driver-loader` into this repo would be a good idea, *but* it wouldn't really be a standalone script then (we don't want to hardcode the Falco driver repo for all consumers).

Maybe it would work to make two scripts: one to build the driver and another one to download a prebuilt driver (with the logic to determine the config hash etc.). Consumers could then wrap these two however they want and provide the prebuilt drivers from wherever they want.

Not sure about the docker-entrypoint script. There will have to be some per-consumer logic (at the very minimum, we're starting different processes in the end) so maybe it would work as a shell library to be sourced into the consumers' entry point scripts.

Reply:
Yes, it doesn't have to be exactly that `falco-driver-loader` file, but something that does something similar.  Either builds the ebpf/kernel probe from source or downloads from a pre-built `scap` probe location - that could be configurable if a consumer wants to use its own pre-built probe repo.  

All the docker-entrypoint.sh does is mount the kernel header/src directories to /host/ in the container and then calls the aforementioned driver loader.   When porting the sysflow collector over from the sysdig libs to the falco libs, I still needed to link to the falco project proper for three main reasons:
1.) Copy over the falco-driver-loader file to automatically build or download the prebuilt probes.
2.) Copy over the docker-entrypoint.sh to setup the environment so the probe could be built properly and the driver loader called,
3.) the falco-libs patch file which renamed SYSDIG_HOST_ROOT to HOST_ROOT and .sysdig to .falco

Consumers could wrap this core functionality with whatever added features they wanted.  Just a thought.

Reply:
Will have to think about it, but I think it could live as a separate PR/proposal, independent of this one.

I like the idea, I'd just like to avoid scope creep

Reply:
> 3.) the falco-libs patch file which renamed SYSDIG_HOST_ROOT to HOST_ROOT and .sysdig to .falco

This point and the defaulted canonical name should be addressed by https://github.com/falcosecurity/libs/pull/41 (N.B. the PR still use `falco` as default, but I will be happy to change it to `scap` once this proposal has been approved)

Regarding other points, I agree with @gnosek : another independent proposal would be better.

Just my 2 cents